crates/node/p2p/src/net/broadcast.rs

```diff
@@ -34,17 +39,40 @@ impl Broadcast {
     /// If an error occurs, the function will exit early, maintaining the invariant that unsafe
     /// blocks are held in order.
     pub fn broadcast(&mut self) {
-        loop {
-            if self.buffer.is_empty() {
-                break;
-            }
-            if let Some(block) = self.buffer.front() {
-                if let Err(e) = self.channel.send(block.clone()) {
-                    trace!("Failed to send unsafe block through channel: {:?}", e);
+        if self.broadcasting.lock().map_or(true, |v| *v) {
+            trace!(target: "net", "Broadcasting in flight, skipping");
+            return;
+        }
+        if let Ok(mut broadcasting) = self.broadcasting.lock() {
+            *broadcasting = true;
+        }
+        let flag = Arc::clone(&self.broadcasting);
+        let channel = self.channel.clone();
+        let mut drained = self.buffer.drain(..).collect::<VecDeque<_>>();
```

Comment:
Can we just `mem::take` here instead? No need to iterate and drain if we're doing the full thing.

Reply:
Good catch thanks